File: solutions/0098_ValidateBinarySearchTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root) {
    function helper(root, lower_limit, upper_limit){
        if(!root){
            return true;
        }
        
        if(root.val <= lower_limit || root.val >= upper_limit){
            return false;
        }
        
        let left_side = helper(root.left,lower_limit,root.val) // elements from left subtree should be less than parent node
        if (!left_side){
            return false;
        }
        
        let right_side = helper(root.right,root.val,upper_limit) // elements from right subree should be greater than parent node
        if(!right_side){
            return false;
        }
        
        return true;
    }
    return helper(root,Number.NEGATIVE_INFINITY,Number.POSITIVE_INFINITY)
};

File: solutions/0098_ValidateBinarySearchTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        return self.helper(root)
    
    def helper(self, root, low = float('-inf'), up = float('inf')):
        if not root:
            return True
        
        if root.val <= low or root.val >= up:
            return False
        
        # left child.val should be less than root.val
        if not self.helper(root.left, low, root.val):
            return False
        
        # right child.val should be more than root.val
        if not self.helper(root.right, root.val, up):
            return False
        
        return True